packages/core/src/molecule/codec.ts

```diff
@@ -0,0 +1,91 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { bytesFrom, BytesLike } from "../bytes/index.js";
+import {
+  Num,
+  numBeFromBytes,
+  numBeToBytes,
+  numFromBytes,
+  numToBytes,
+} from "../num/index.js";
+
+export interface Codec<Encodable, Decodable = BytesLike> {
+  byteLength?: number;
+  encode: (encodable: Encodable) => Decodable;
+  decode: (decodable: Decodable) => Encodable;
+}
+
+export type EncodableType<T extends Codec<any, any>> =
+  T extends Codec<infer Encodable, any> ? Encodable : never;
+export type DecodableType<T extends Codec<any, any>> =
+  T extends Codec<any, infer Decodable> ? Decodable : never;
+
+/**
+ * Create a codec to deal with bytes-like data.
+ * @param codec
+ */
+export function bytes<Encodable>(codec: Codec<Encodable>): Codec<Encodable> {
```

Comment:
It seems that `bytes` and `fixedBytes` are just for type conversion and checking?